funsor/gaussian.py

```diff
@@ -380,42 +382,74 @@ def eager_subs(self, subs):
+            values[k] = value.expand(batch_shape + value.shape[-1:])
 
+        # Try to perform a complete substitution of all real variables, resulting in a Tensor.
+        if all(k in real_subs for k, d in self.inputs.items() if d.dtype == 'real'):
             # Form the concatenated value.
-            offsets, event_size = _compute_offsets(self.inputs)
             value = BlockVector(batch_shape + (event_size,))
-            for k, value_k in zip(real_subs, values):
-                offset = offsets[k]
-                value_k = value_k.reshape(value_k.shape[:batch_dim] + (-1,))
-                if not torch._C._get_tracing_state():
-                    assert value_k.size(-1) == self.inputs[k].num_elements
-                value_k = value_k.expand(batch_shape + value_k.shape[-1:])
-                value[..., offset: offset + self.inputs[k].num_elements] = value_k
+            for k, i in slices:
+                if k in values:
+                    value[..., i] = values[k]
             value = value.as_tensor()
 
             # Evaluate the non-normalized log density.
             result = -0.5 * _vmv(precision, value - loc)
-            result = Tensor(result, inputs)
+            result = Tensor(result, int_inputs)
             assert result.output == reals()
             return Subs(result, lazy_subs)
 
         # Perform a partial substution of a subset of real variables, resulting in a Joint.
-        # See "The Matrix Cookbook" (November 15, 2012) ss. 8.1.3 eq. 353.
-        # http://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf
-        raise NotImplementedError('TODO implement partial substitution of real variables')
+        # We split real inputs into two sets: a for the preserved and b for the substituted.
+        raise NotImplementedError('TODO')
+        b = frozenset(k for k, v in real_subs.items())
+        a = frozenset(k for k, d in self.inputs.items() if d.dtype == 'real' and k not in b)
+        loc_a = torch.cat([loc[..., i] for k, i in slices if k in a], dim=-1)
+        diff_b = torch.cat([values[k] - loc[..., i] for k, i in slices if k in b], dim=-1)
+        prec_aa = torch.cat([torch.cat([
+            precision[..., i1, i2]
+            for k2, i2 in slices if k2 in a], dim=-1)
+            for k1, i1 in slices if k1 in a], dim=-2)
+        prec_ab = torch.cat([torch.cat([
+            precision[..., i1, i2]
+            for k2, i2 in slices if k2 in b], dim=-1)
+            for k1, i1 in slices if k1 in a], dim=-2)
+        prec_bb = torch.cat([torch.cat([
+            precision[..., i1, i2]
+            for k2, i2 in slices if k2 in b], dim=-1)
+            for k1, i1 in slices if k1 in b], dim=-2)
+        prec_ab_diff_b = _mv(prec_ab, diff_b)
+        loc = loc_a + (prec_ab_diff_b.unsqueeze(-1)
+                                     .cholesky_solve(prec_aa.cholesky())
+                                     .squeeze(-1))
+        log_scale = _vv(loc_a, prec_ab_diff_b) - 0.5 * _vmv(prec_bb, diff_b)
```

Comment:
I think the first term `_vv(loc_a, prec_ab_diff_b)` -> `0.5 * _vv(loc - loc_a, prec_ab_diff_b)`, where `loc - loc_a` is obtained from the above computation.